appstudio-controller/controllers/appstudio.redhat.com/deploymenttargetbinder_controller.go

```diff
@@ -0,0 +1,577 @@
+*/
+
+package appstudioredhatcom
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	applicationv1alpha1 "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	sharedutil "github.com/redhat-appstudio/managed-gitops/backend-shared/util"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// DeploymentTargetClaimReconciler reconciles a DeploymentTargetClaim object
+type DeploymentTargetClaimReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargetclaims,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargetclaims/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargetclaims/finalizers,verbs=update
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargets,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=appstudio.redhat.com,resources=deploymenttargets/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the DeploymentTargetClaim object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *DeploymentTargetClaimReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx).WithValues("request", req)
+
+	dtc := applicationv1alpha1.DeploymentTargetClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(&dtc), &dtc); err != nil {
+		// Don't requeue if the requested object is not found/deleted.
+		if apierr.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Add the deletion finalizer if it is absent.
+	if addFinalizer(&dtc, sharedutil.FinalizerBinder) {
+		if err := r.Client.Update(ctx, &dtc); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer %s to DeploymentTargetClaim %s in namespace %s: %v", sharedutil.FinalizerBinder, dtc.Name, dtc.Namespace, err)
+		}
+	}
```

Comment:
Yeah. According to this [finalizer table](https://github.com/redhat-appstudio/book/blob/main/ADR/0008-environment-provisioning.md#finalizer-details), the provisioner will add the finalizer to the `DeploymentTarget` . Although the table doesn't mention any finalizer for `DeploymentTargetClaim`, the [DT and DTC lifecycle workflow ](https://github.com/redhat-appstudio/book/blob/main/ADR/0008-environment-provisioning.md#deploymenttargetdt-and-deploymenttargetclaimdtc-lifecycle) implies that the binding controller will add the finalizer to DTC. 

From the lifecycle workflow (after the DTC marked for deletion): 
```
Deletion is delayed by a finalizer that was set by the binding controller
```

Also I couldn't think of any scenario for controlling the DT finalizer in the binding controller since the provisioner will take care of handling DT deletion. That is what I understood from the [Deprovision Loop](https://github.com/redhat-appstudio/book/blob/main/ADR/0008-environment-provisioning.md#deprovision-loop-dt-reconciliation)

Please do let me know if you could think of any usecase for managing the DT finalizer in the binding controller.